### ISSUE TITLE
ESMFold Embedding

### ISSUE THREAD

**Issue by alishbaimran** [alishbaimran/protein-bi-encoder#1]
My understanding of the current pipeline here:
![image](https://github.com/alishbaimran/protein-bi-encoder/assets/44557946/1db2718c-ab35-4f25-9dfd-27ab1dc58195)

- ESM2 layer output: esm_s is a tensor containing the outputs from multiple layers of the ESM2 model. Specifically, the model takes in the entire protein sequence. For each amino acid in the sequence, the model generates embeddings at multiple layers. The result is a tensor (esm_s) containing embeddings for each amino acid at each layer of the model.
- The shape of esm_s is [batch_size, sequence_length, num_layers, embedding_dim], batch_size is the # of protein sequences in the batch, sequence_length is the length of each protein sequence, num_layers is the number of layers in the ESM2 model, embedding_dim is the dimension of the embeddings produced by each layer.
- The learned weights are contained in esm_s_combine which is a tensor of shape [num_layers + 1]
- The softmax-normalized weights are used to perform a weighted sum of the layer outputs in esm_s. So each layer's output is combined into a single representation for each amino acid in the sequence. 

```
        esm_s, esm_z = self._compute_language_model_representations(esmaa)
        esm_s = esm_s.to(self.esm_s_combine.dtype)
        esm_s = esm_s.detach()

        # === preprocessing ===
        esm_s = (self.esm_s_combine.softmax(0).unsqueeze(0) @ esm_s).squeeze(2)
```

- esm_s has a final shape of [batch_size, sequence_length, embedding_dim] here. 
- Embeddings passed through MLP: s_s_0
- For each amino acid in the sequence, the sequence-level embeddings (s_s_0) is combined with the corresponding amino acid embedding using element-wise addition
- These embeddings are fed into the ESMFold structure module (Folding Trunk), which predicts the 3D structure of the protein.

So the **raw output from ESM2 here is (esm_s) which represent the outputs from each layer of the ESM2 model**. "x" are the processed embeddings after combining layer outputs, refining with an MLP, and adding amino acid embeddings.

**Goal**:
Replace the current ESM2 output used in the bi-encoder with the processed embeddings (x) from the ESMFold model.

Tasks:
- generating processed embeddings (x) using the ESMFold model instead of raw ESM2 outputs.
- updating the bi-encoder model to use these processed embeddings for learning the contrastive relationship between protein sequences and their corresponding natural language descriptions.
